Sources/PostgresNIO/New/PostgresChannelHandler.swift

```diff
@@ -317,53 +323,74 @@ final class PostgresChannelHandler: ChannelDuplexHandler {
 
     // MARK: Channel handler actions
     
-    private func run(_ action: ConnectionStateMachine.ConnectionAction, with context: ChannelHandlerContext) {
+    private func run(
+        _ action: ConnectionStateMachine.ConnectionAction,
+        with context: ChannelHandlerContext,
+        promise runPromise: EventLoopPromise<Void>?
+    ) {
         self.logger.trace("Run action", metadata: [.connectionAction: "\(action)"])
         
         switch action {
         case .establishSSLConnection:
-            self.establishSSLConnection(context: context)
+            self.establishSSLConnection(context: context, promise: runPromise)
         case .read:
             context.read()
+            runPromise?.succeed(())
         case .wait:
-            break
+            runPromise?.succeed(())
         case .sendStartupMessage(let authContext):
             self.encoder.startup(user: authContext.username, database: authContext.database, options: authContext.additionalParameters)
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .sendSSLRequest:
             self.encoder.ssl()
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .sendPasswordMessage(let mode, let authContext):
-            self.sendPasswordMessage(mode: mode, authContext: authContext, context: context)
+            self.sendPasswordMessage(mode: mode, authContext: authContext, context: context, promise: runPromise)
         case .sendSaslInitialResponse(let name, let initialResponse):
             self.encoder.saslInitialResponse(mechanism: name, bytes: initialResponse)
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .sendSaslResponse(let bytes):
             self.encoder.saslResponse(bytes)
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .closeConnectionAndCleanup(let cleanupContext):
             self.closeConnectionAndCleanup(cleanupContext, context: context)
+            if let promise = cleanupContext.closePromise {
+                promise.futureResult.cascade(to: runPromise)
+            } else {
+                runPromise?.succeed(())
+            }
```

Comment:
If the cleanup context has a promise in it, cascade the result of that to `runPromise`.
Other wise succeed `runPromise`.